Lab1_Operationalizing_Pytorch_with_Mlflow/LAB.md

```diff
@@ -29,5 +24,101 @@ itself is organized into subfolders according to which components it belongs to;
+[Continuous Integration](https://en.wikipedia.org/wiki/Continuous_integration) (CI), 
+[Continuous Delivery](https://en.wikipedia.org/wiki/Continuous_delivery) (CD) and 
+[Continuous Traning](https://cloud.google.com/solutions/machine-learning/mlops-continuous-delivery-and-automation-pipelines-in-machine-learning) (CT). 
+The following figure depicts the proposed architecture.
+
+![Proposed architecture](imgs/arch.svg)
+
+**Nodes** are bare-metal or virtual machines that host a part of the pipeline process. **ML components** 
+are ML-based code that should be trained, tested and deployed by end-to-end pipeline.They must be 
+packaged as [MLFlow projects](https://mlflow.org/docs/latest/projects.html). **Training nodes**
+host the training process which runs in **Training containers**, running dockerized images built 
+with MLFlow and necessary libs to train ML components. Each ML component defines its needs, that is, 
+libs and external dependencies. To ensure traceability, provenance and model configuration management, 
+one may track hyperparameters, training steps, test results, extra data and the model itself into 
+_MLFlow Tracking Server_ hosted in the **Tracking Node**. All these items, metadata and trained models, 
+are stored in external resources: RDBMS and Cloud Storage Services. Trained models are embedded into 
+dockerized **Serving Instances**, containers running [Torchserve](https://pytorch.org/serve/) providing 
+web access to the models. Serving instances are hosted by **Serving Nodes**.
+
+In order to provide feedback loop, a **Monitor Node** checks the status and performance meters, issuing 
+new training cycles or a full rebuild process. Some structures traverses through Development and Production perimeters. Serving, Training and Monitor nodes are key elements that communicate between those zones. 
+**Control or Operator Machine** is the VM used by Configuration Manager or Operation Engineer to follow up
+the process or issue commands to the nodes in the pipeline. This process may run through many different 
+perimeters, however this solution comprises _Development_ and _Production_ stages only.
+
+### Architectural Components and Decisions
+
+In this section, we record the (most relevant) design decisions, components selection and the rationale around them.
+
+| # | Subject | Description | Affected Items | Solution | Rationale |
+|---|---------|-------------|----------------|----------|-----------|
+| 1 | Computing | ML needs processing power as well as GPU. Our architecure relies on distributed elements, thus needing a reasonable number of physical or virtual machines (some with GPU-power) | All Nodes and Machines | GCP Compute Service | It provides $ 300,00 of (enough) free services and we can also spin some gpu-powered machines |
+| 2 | Durable Data Storage | Every serious solution needs to store durable data | Tracking Server | SQLite | Simplest and Fastest Relational Database tha fits our needs |
```

Comment:
This is a postgres server hosted in google cloud

Reply:
Needed to be permanent, since a sqlite database could go down with the VM and lose all experiment data

Reply:
"needed"

Reply:
Much better. So we should change to the service name. I guess it's Google Cloud SQL.

Reply:
```suggestion
| 2 | Durable Data Storage | Every serious solution needs to store durable data | Tracking Server | Google Cloud SQL | A Postgres Database hosted by Google Cloud SQL is reliable and fast enough for our needs  |
```